fem/kernels/fespace.cpp

```diff
@@ -0,0 +1,88 @@
+// Copyright (c) 2010, Lawrence Livermore National Security, LLC. Produced at
+// the Lawrence Livermore National Laboratory. LLNL-CODE-443211. All Rights
+// reserved. See file COPYRIGHT for details.
+//
+// This file is part of the MFEM library. For more information and source code
+// availability see http://mfem.org.
+//
+// MFEM is free software; you can redistribute it and/or modify it under the
+// terms of the GNU Lesser General Public License (as published by the Free
+// Software Foundation) version 2.1 dated February 1999.
+
+#include "../../general/okina.hpp"
+
+namespace mfem
+{
+namespace kernels
+{
+namespace fem
+{
+
+// *****************************************************************************
+void L2E(const int NUM_VDIM,
+         const bool VDIM_ORDERING,
+         const int globalEntries,
+         const int localEntries,
+         const int* __restrict offsets,
+         const int* __restrict indices,
+         const double* __restrict globalX,
+         double* __restrict localX)
+{
+   GET_CONST_PTR_T(offsets,int);
+   GET_CONST_PTR_T(indices,int);
+   GET_CONST_PTR(globalX);
+   GET_PTR(localX);
+   MFEM_FORALL(i, globalEntries,
+   {
+      const int offset = d_offsets[i];
+      const int nextOffset = d_offsets[i+1];
+      for (int v = 0; v < NUM_VDIM; ++v)
+      {
+         const int g_offset = ijNMt(v,i,NUM_VDIM,globalEntries,VDIM_ORDERING);
```

Comment:
I really think that instead of having these `ijNMt`, `ijklNM` obscure functions, we should start thinking about how we could have nice multi-dimension arrays with an associated Layout, I really like the Raja approach: https://raja.readthedocs.io/en/master/feature/view.html#raja-layout